velox/functions/sparksql/DateTimeFunctions.h

```diff
@@ -372,4 +372,78 @@ struct AddMonthsFunction {
     result = daysSinceEpoch;
   }
 };
+
+static const folly::F14FastMap<std::string, int8_t> kDayOfWeekNames{
+    {"th", 0},       {"fr", 1},     {"sa", 2},       {"su", 3},
+    {"mo", 4},       {"tu", 5},     {"we", 6},       {"thu", 0},
+    {"fri", 1},      {"sat", 2},    {"sun", 3},      {"mon", 4},
+    {"tue", 5},      {"wed", 6},    {"thursday", 0}, {"friday", 1},
+    {"saturday", 2}, {"sunday", 3}, {"monday", 4},   {"tuesday", 5},
+    {"wednesday", 6}};
+
+template <typename T>
+struct NextDayFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& config,
+      const arg_type<Date>* /*startDate*/,
+      const arg_type<Varchar>* dayOfWeek) {
+    if (dayOfWeek != nullptr) {
+      try {
+        weekDay_ = getDayOfWeekFromString(
+            std::string_view(dayOfWeek->data(), dayOfWeek->size()));
+      } catch (const VeloxUserError&) {
+        invalidFormat_ = true;
+      }
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE int8_t
+  getDayOfWeekFromString(std::string_view dayOfWeek) {
+    VELOX_USER_CHECK(!dayOfWeek.empty(), "Day of week should not be empty")
+    std::string lowerDayOfWeek =
+        boost::algorithm::to_lower_copy(std::string(dayOfWeek));
+    auto it = kDayOfWeekNames.find(lowerDayOfWeek);
+    VELOX_USER_CHECK(
+        it != kDayOfWeekNames.end(),
+        "Illegal input for day of week: {}",
+        dayOfWeek);
+    return it->second;
+  }
+
+  FOLLY_ALWAYS_INLINE int64_t getNextDate(int64_t startDay, int8_t dayOfWeek) {
+    return startDay + 1 + ((dayOfWeek - 1 - startDay) % 7 + 7) % 7;
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Date>& result,
+      const arg_type<Date>& startDate,
+      const arg_type<Varchar>& dayOfWeek) {
+    VELOX_USER_CHECK(
+        !invalidFormat_, "Illegal input for day of week: {}", dayOfWeek)
+    if (!weekDay_.has_value()) {
+      weekDay_ = getDayOfWeekFromString(
```

Comment:
There is a bug here. weekDay_ is a member variable which is supposed to store the constant value, but you are updating it with non-constant value.

Reply:
@mbasmanova  Yes, here it is still necessary to use isConstFormat_ and weekDay_ in combination. Using only one std::optional<int8_t> variable will introduce the bug.

Reply:
@PHILO-HE here.